File: lib/scanner/plugins/servers/python.rb
# frozen_string_literal: true

module Yawast
  module Scanner
    module Plugins
      module Servers
        class Python
          def self.check_banner(banner)
            Yawast::Shared::Output.log_hash 'vulnerabilities',
                                            'python_version_exposed',
                                            {vulnerable: false, version: nil}

            # don't bother if this doesn't include Python
            return unless banner.include? 'Python/'

            Yawast::Utilities.puts_warn "Python Version: #{banner}"
            puts ''

            Yawast::Shared::Output.log_hash 'vulnerabilities',
                                            'python_version_exposed',
                                            {vulnerable: true, version: banner}
          end
        end
      end
    end
  end
end

File: yawast/scanner/plugins/http/servers/python.py
from typing import List

from yawast.reporting.enums import Vulnerabilities
from yawast.scanner.plugins.result import Result


def check_banner(banner: str, raw: str, url: str) -> List[Result]:
    if not banner.startswith("Python/"):
        return []

    results = []

    # we've got a version
    results.append(
        Result(
            f"Python Version Exposed: {banner}",
            Vulnerabilities.HTTP_BANNER_PYTHON_VERSION,
            url,
            raw,
        )
    )

    return results
